dwst/scripts/test/particles.js

```diff
@@ -213,12 +213,153 @@ describe('particles module', () => {
+    });
+    it('should throw correct InvalidParticles for missing expression open and close', () => {
+      expect(() => {
+        return parseParticles('$foo()');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '$foo()',
+        expected: '{',
+        remainder: 'foo()',
+        errorPosition: '$'.length,
+      });
+    });
+    it('should throw correct InvalidParticles for missing argument', () => {
+      expect(() => {
+        return parseParticles('${foo(123,)}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(123,)}',
+        expected: 'argument',
+        remainder: ')}',
+        errorPosition: '${foo(123,'.length,
+      });
+    });
+    it('should throw correct InvalidParticles for missing first argument', () => {
+      expect(() => {
+        return parseParticles('${foo(,456)}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(,456)}',
+        expected: 'argument',
+        remainder: ',456)}',
+        errorPosition: '${foo('.length,
+      });
+    });
+    it('should throw correct InvalidParticles for missing argument list', () => {
+      expect(() => {
+        return parseParticles('${foo}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo}',
+        expected: '(',
+        remainder: '}',
+        errorPosition: '${foo'.length,
+      });
+    });
+    it('should throw correct InvalidParticles for unterminated expression with empty argument list missing close', () => {
+      expect(() => {
+        return parseParticles('${foo(}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(}',
+        expected: 'argument or )',
+        remainder: '}',
+        errorPosition: '${foo('.length,
+      });
+    });
+    it('should throw correct InvalidParticles for unterminated expression with nonempty argument list missing close', () => {
+      expect(() => {
+        return parseParticles('${foo(123');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(123',
+        expected: ', or )',
+        remainder: '',
+        errorPosition: '${foo(123'.length,
+      });
+    });
+    it('should throw correct InvalidParticles for unterminated expression with nonempty argument list missing argument and close', () => {
+      expect(() => {
+        return parseParticles('${foo(123,');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(123,',
+        expected: 'argument',
+        remainder: '',
+        errorPosition: '${foo(123,'.length,
+      });
+    });
+    it('should throw correct InvalidParticles for unterminated expression with nonempty multivalue argument list missing close', () => {
+      expect(() => {
+        return parseParticles('${foo(123,456');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(123,456',
+        expected: ')',
```

Comment:
maybe expected should be `'")"'` to indicate that we are *literally* expecting the character `)`

Reply:
Apparently instructions can never have more than two arguments? That is probably fine at the moment though. However, I did not realize that this was the case.